app/models/ont/flowcell.rb

```diff
@@ -5,18 +5,106 @@ module Ont
   class Flowcell < ApplicationRecord
     include Uuidable
 
+    # flowcell_ids have 3 letters followed by 3 numbers
+    FLOWCELL_ID_FORMAT = 'ABC%03d'
+
     # Run has many of these flowcells up to the maximum number for the instrument.
     belongs_to :run, foreign_key: :ont_run_id, inverse_of: :flowcells
 
-    # We assume one-to-one relationship with pool at the moment.
-    belongs_to :pool, foreign_key: :ont_pool_id, inverse_of: :flowcell
+    # We assume one-to-one relationship with pool. We make it optional here to
+    # remove the default presence validation to customise the validations below.
+    belongs_to :pool, foreign_key: :ont_pool_id, inverse_of: :flowcell, optional: true
 
     delegate :requests, :libraries, to: :pool
 
-    # Validate the position to be a positive integer.
-    validates :position,
-              presence: true,
-              numericality: { greater_than_or_equal_to: 1 }
-    validates :flowcell_id, presence: true
+    # flowcell position validations
+    # position must exist.
+    # position must be an integer.
+    # position must be in range between one and the max number of the instrument.
+    # position must be unique among the others in the same run. This scoped
+    # position uniqueness within the run is validated by run because the scoped
+    # uniqueness validation does not work properly when run does not have an id
+    # and it accepts nested attributes to autosave its flowcells.
+    # Numericality validations are listed separately to avoid duplicated
+    # validation keys.
+
+    validates :position, presence: true
+    validates :position, numericality: { only_integer: true }
+    validates :position, numericality: {
+      greater_than_or_equal_to: 1,
+      less_than_or_equal_to: :max_number_of_flowcells,
+      if: :max_number_of_flowcells,
+      message: :position_out_of_range
+    }
+
+    # pool validations
+    # pool must exist. The error message is given as "is unknown" because of
+    # how the UI shows the message. The UI gets a tube barcode and finds the
+    # corresponding pool. If it can't find the pool, it sends a nil
+    # ont_pool_id . When it shows the error message, the invalid tube barcode
+    # entered by the user is still shown in the page. If the error message was
+    # "is required" or "is missing", it would be confusing.
+    # We do not validate ont_pool_id separately as the presence of the pool
+    # covers it.
+    # We used generate_message to be able to pass options from this object.
```

Comment:
The UI needs to tell user at which position there is an error in different formats for different instrument runs. Standard errors were not enough. We need to put flowcell.position_display into the message.

In case or "errors.add" used in custom validation methods, it was relatively easy to pass the flowcell.position_display into the locale string, as it does the searching through the namespace in the locale file for the object.

However, I had to use generate_message to build the message when we are using built-in validations because I could not put flowcell.position_display into the scope of validation while handling the locale string. 

It also saves me from handling the locale string namespace handling. Direct uses of I18n was very messy. 

The errors.add method also use generate_message.